gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactorJobRunner.java

```diff
@@ -397,4 +441,103 @@ public Status status() {
   public int compareTo(MRCompactorJobRunner o) {
     return Double.compare(o.priority, this.priority);
   }
+
+  private List<Path> getCumulativeLateFilePaths(Path lateDataDir) throws FileNotFoundException, IOException {
+    if (!this.fs.exists(lateDataDir)) {
+      return Lists.newArrayList();
+    }
+    List<Path> paths = Lists.newArrayList();
+    for (FileStatus fileStatus : FileListUtils.listFilesRecursively(fs, lateDataDir, new PathFilter() {
+      @Override
+      public boolean accept(Path path) {
+        for (String validExtention : getApplicableFileExtensions()) {
+          if (path.getName().endsWith(validExtention)) {
+            return true;
+          }
+        }
+        return false;
+      }
+    })) {
+      paths.add(fileStatus.getPath());
+    }
+    return paths;
+  }
+
+  /**
+   * Submit an event reporting new and cumulative late record counts.
+   */
+  private void submitLateRecordCountsEvent(List<Path> newLateFilePaths, Path lateDataDir) {
+    try {
+      if (newLateRecordCountProvider instanceof RecordCountFromFileName) {
```

Comment:
There is some breakage of abstraction here if your have a generic interface like FsRecordCountProvider but then you have to check for specific implementations like RecordCountFromFileName).